File: NPC.mwherman2000.NEP5TokenPython/NPC.mwherman2000.NEP5Token.ManualPContract/NEP5LedgerEntryL2Persistable.py
from NeoTrace import NeoTraceRuntime
from NeoEntityModel import NeoEntityModel
from boa.interop.Neo.Storage import GetContext, Get, Put, Delete
from boa.interop.BigInteger import FromBytes, ToByteArray
from boa.code.builtins import concat

class NEP5LedgerEntry(NeoTraceRuntime):

    __className = "NEP5LedgerEntry"
    __bClassName = bytes(__className)
    __sTimestamp = "Timestamp"
    __bTimestamp = bytes(__sTimestamp)
    __sDecription = "Decription"
    __bDecription = bytes(__sDecription)
    __sDebitCreditAmount = "DebitCreditAmount"
    __bDebitCreditAmount = bytes(__sDebitCreditAmount)
    __sBalance = "Balance"
    __bBalance = bytes(__sBalance)
    __sSTA = "_STA"
    __bSTA = bytes(__sSTA)
    __sEXT = "_EXT"
    __bEXT = bytes(__sEXT)
    __classKeyTag = concat(concat("/#", __className),".")
    __bclassKeyTag = bytes(__classKeyTag)

    def IsMissing(e: NEP5LedgerEntry):
        return (e.__state == NeoEntityModel.MISSING)

    def Missing():
        e.__timestamp = 0
        e.__decription = ""
        e.__debitCreditAmount = 0
        e.__balance = 0
        e.__state = NeoEntityModel.MISSING
        if NeoTrace.RUNTIME:
            TraceRuntime("Missing().NEP5LedgerEntry", e)
        return e

    def Put(e: NEP5LedgerEntry, key: bytes):
        if len(key) == 0:
            return False
        ctx = GetContext()
        __bkeyTag = concat(key, self.__bclassKeyTag)

        e.__state = NeoEntityModel.PUTTED
        Put(ctx, concat(__bkeyTag, self.__bSTA), e.__state)
        Put(ctx, concat(__bkeyTag, self.__bTimestamp), e.__timestamp)
        Put(ctx, concat(__bkeyTag, self.__bDecription), e.__decription)
        Put(ctx, concat(__bkeyTag, self.__bDebitCreditAmount), e.__debitCreditAmount)
        Put(ctx, concat(__bkeyTag, self.__bBalance), e.__balance)
        if NeoTrace.RUNTIME:
            LogExt("Put(bkey).NEP5LedgerEntry", e)
        return True

    def Put(e: NEP5LedgerEntry, key: str):
        if len(key) == 0:
            return False
        if NeoTrace.RUNTIME:
            LogExt("Put(skey).NEP5LedgerEntry", e)
        ctx = GetContext()
        __skeyTag = concat(key, self.__classKeyTag)
        if NeoTrace.RUNTIME:
            TraceRuntime("Put(skey).__skeyTag", __skeyTag)
        e.__state = NeoEntityModel.PUTTED
        if NeoTrace.RUNTIME:
            TraceRuntime("Put(skey).bis", e.__state)
        Put(ctx, concat(__skeyTag, self.__sSTA), e.__state)
        Put(ctx, concat(__skeyTag, self.__sTimestamp), e.__timestamp)
        Put(ctx, concat(__skeyTag, self.__sDecription), e.__decription)
        Put(ctx, concat(__skeyTag, self.__sDebitCreditAmount), e.__debitCreditAmount)
        Put(ctx, concat(__skeyTag, self.__sBalance), e.__balance)
        if NeoTrace.RUNTIME:
            LogExt("Put(bkey).NEP5LedgerEntry", e)
        return True

    def Get(key: bytes):
        if len(key) == 0:
            return None
        ctx = GetContext()
        __bkeyTag = concat(key, self.__bclassKeyTag)
        bsta = Get(ctx, concat(__bkeyTag, __bSTA))
        if NeoTrace.RUNTIME:
            TraceRuntime("Get(bkey).bsta", len(bsta), bsta)
        if len(bsta) == 0:
            e = NEP5LedgerEntry.Missing()
        else:
            ista = FromBytes(bsta)
            e = NEP5LedgerEntry()
            e.__timestamp = Get(ctx, concat(__bkeyTag, self.__bTimestamp))
            e.__decription = Get(ctx, concat(__bkeyTag, self.__bDecription))
            e.__debitCreditAmount = Get(ctx, concat(__bkeyTag, self.__bDebitCreditAmount))
            e.__balance = Get(ctx, concat(__bkeyTag, self.__bBalance))
            e.__state = NeoEntityModel.GETTED
        if NeoTrace.RUNTIME:
            LogExt("Get(bkey).NEP5LedgerEntry", e)

    def Get(key: str):
        if len(key) == 0:
            return None
        ctx = GetContext()
        __skeyTag = concat(key, self.__classKeyTag)
        bsta = Get(ctx, concat(__skeyTag, __sSTA))
        if NeoTrace.RUNTIME:
            TraceRuntime("Get(bkey).bsta", len(bsta), bsta)
        if len(bsta) == 0:
            e = NEP5LedgerEntry.Missing()
        else:
            ista = FromBytes(bsta)
            e = NEP5LedgerEntry()
            e.__timestamp = Get(ctx, concat(__skeyTag, self.__sTimestamp))
            e.__decription = Get(ctx, concat(__skeyTag, self.__sDecription))
            e.__debitCreditAmount = Get(ctx, concat(__skeyTag, self.__sDebitCreditAmount))
            e.__balance = Get(ctx, concat(__skeyTag, self.__sBalance))
            e.__state = NeoEntityModel.GETTED
        if NeoTrace.RUNTIME:
            LogExt("Get(bkey).NEP5LedgerEntry", e)
        return e

File: NPC.mwherman2000.NEP5TokenPython/NPC.mwherman2000.NEP5Token.Contract/NEP5LedgerEntryL2Persistable.cs
using NPC.Runtime;
using Neo.SmartContract.Framework;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

/// <summary>
/// NPC.mwherman2000.NEP5Token.Contract.NEP5LedgerEntry - Level 2 Persistable
///
/// Processed:       2018-04-21 10:49:17 PM by npcc - NEO Persistable Classes (NPC) Platform 2.1 Compiler v2.1.0.38328
/// NPC Project:     https://github.com/mwherman2000/neo-npcc2/blob/master/README.md
/// NPC Lead:        Michael Herman (Toronto) (mwherman@parallelspace.net)
/// </summary>

namespace NPC.mwherman2000.NEP5Token.Contract
{
    public partial class NEP5LedgerEntry : NeoTraceRuntime /* Level 2 Persistable */
    {
        // Class name and property names
        private const string _className = "NEP5LedgerEntry";
        private static readonly byte[] _bClassName = _className.AsByteArray();

        private const string _sTimestamp = "Timestamp"; // Template: NPCLevel2AFieldConsts_cs.txt
        private static readonly byte[] _bTimestamp = Helper.AsByteArray(_sTimestamp);
        private const string _sDecription = "Decription"; // Template: NPCLevel2AFieldConsts_cs.txt
        private static readonly byte[] _bDecription = Helper.AsByteArray(_sDecription);
        private const string _sDebitCreditAmount = "DebitCreditAmount"; // Template: NPCLevel2AFieldConsts_cs.txt
        private static readonly byte[] _bDebitCreditAmount = Helper.AsByteArray(_sDebitCreditAmount);
        private const string _sBalance = "Balance"; // Template: NPCLevel2AFieldConsts_cs.txt
        private static readonly byte[] _bBalance = Helper.AsByteArray(_sBalance);
        private const string _sSTA = "_STA"; // Template: NPCLevel2BMissing_cs.txt
        private static readonly byte[] _bSTA = Helper.AsByteArray(_sSTA);

        private const string _sEXT = "_EXT";
        private static readonly byte[] _bEXT = Helper.AsByteArray(_sEXT);
        
        // Internal fields
        private const string _classKeyTag = "/#" + _className + ".";
        private static readonly byte[] _bclassKeyTag = Helper.AsByteArray(_classKeyTag);
 
        // Persistable methods
        public static bool IsMissing(NEP5LedgerEntry e)
        {
            return (e._state == NeoEntityModel.EntityState.MISSING);
        }

        public static NEP5LedgerEntry Missing()
        {
            NEP5LedgerEntry e = new NEP5LedgerEntry();
            e._timestamp = 0; e._decription = ""; e._debitCreditAmount = 0; e._balance = 0; 
            e._state = NeoEntityModel.EntityState.MISSING;
            if (NeoTrace.RUNTIME) LogExt("Missing().NEP5LedgerEntry", e);
            return e;
        }

        public static bool Put(NEP5LedgerEntry e, byte[] key)
        {
            if (key.Length == 0) return false;

            Neo.SmartContract.Framework.Services.Neo.StorageContext ctx = Neo.SmartContract.Framework.Services.Neo.Storage.CurrentContext;
            byte[] _bkeyTag = Helper.Concat(key, _bclassKeyTag);

            // no readonly fields byte[] bsta = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, Helper.Concat(_bkeyTag, _bSTA));
            // no readonly fields if (NeoTrace.RUNTIME) TraceRuntime("Put(bkey).bsta", bsta.Length, bsta);
            // no readonly fields bool isMissing = false; if (bsta.Length == 0) isMissing = true;

            e._state = NeoEntityModel.EntityState.PUTTED;
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, Helper.Concat(_bkeyTag, _bSTA), e._state.AsBigInteger());

            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, Helper.Concat(_bkeyTag, _bTimestamp), e._timestamp); // Template: NPCLevel2CPut_cs.txt
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, Helper.Concat(_bkeyTag, _bDecription), e._decription); // Template: NPCLevel2CPut_cs.txt
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, Helper.Concat(_bkeyTag, _bDebitCreditAmount), e._debitCreditAmount); // Template: NPCLevel2CPut_cs.txt
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, Helper.Concat(_bkeyTag, _bBalance), e._balance); // Template: NPCLevel2CPut_cs.txt
            if (NeoTrace.RUNTIME) LogExt("Put(bkey).NEP5LedgerEntry", e); // Template: NPCLevel2DPut_cs.txt
            return true;
        }

        public static bool Put(NEP5LedgerEntry e, string key)
        {
            if (key.Length == 0) return false;
            if (NeoTrace.RUNTIME) LogExt("Put(skey).NEP5LedgerEntry", e);

            Neo.SmartContract.Framework.Services.Neo.StorageContext ctx = Neo.SmartContract.Framework.Services.Neo.Storage.CurrentContext;
            string _skeyTag = key + _classKeyTag;
            if (NeoTrace.RUNTIME) TraceRuntime("Put(skey)._skeyTag", _skeyTag);

            // no readonly fields byte[] bsta = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, _skeyTag + _sSTA);
            // no readonly fields if (NeoTrace.RUNTIME) TraceRuntime("Put(skey).bsta", bsta.Length, bsta);
            // no readonly fields bool isMissing = false; if (bsta.Length == 0) isMissing = true;

            e._state = NeoEntityModel.EntityState.PUTTED;
            BigInteger bis = e._state.AsBigInteger();
            if (NeoTrace.RUNTIME) TraceRuntime("Put(skey).bis", bis);
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, _skeyTag + _sSTA, bis);
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, _skeyTag + _sTimestamp, e._timestamp); // Template: NPCLevel2EPut_cs.txt
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, _skeyTag + _sDecription, e._decription); // Template: NPCLevel2EPut_cs.txt
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, _skeyTag + _sDebitCreditAmount, e._debitCreditAmount); // Template: NPCLevel2EPut_cs.txt
            Neo.SmartContract.Framework.Services.Neo.Storage.Put(ctx, _skeyTag + _sBalance, e._balance); // Template: NPCLevel2EPut_cs.txt
            if (NeoTrace.RUNTIME) LogExt("Put(skey).NEP5LedgerEntry", e); // Template: NPCLevel2FGet_cs.txt
            return true;
        }

        public static NEP5LedgerEntry Get(byte[] key)
        {
            if (key.Length == 0) return Null();

            Neo.SmartContract.Framework.Services.Neo.StorageContext ctx = Neo.SmartContract.Framework.Services.Neo.Storage.CurrentContext;
            byte[] _bkeyTag = Helper.Concat(key, _bclassKeyTag);

            NEP5LedgerEntry e;
            byte[] bsta = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, Helper.Concat(_bkeyTag, _bSTA));
            if (NeoTrace.RUNTIME) TraceRuntime("Get(bkey).bsta", bsta.Length, bsta);
            if (bsta.Length == 0)
            {
                e = NEP5LedgerEntry.Missing();
            }
            else // not MISSING
            {
                int ista = (int)bsta.AsBigInteger();
                NeoEntityModel.EntityState sta = (NeoEntityModel.EntityState)ista;
                e = new NEP5LedgerEntry();

                BigInteger Timestamp = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, Helper.Concat(_bkeyTag, _bTimestamp)).AsBigInteger(); //NPCLevel2GGet_cs.txt
                string Decription = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, Helper.Concat(_bkeyTag, _bDecription)).AsString(); //NPCLevel2GGet_cs.txt
                BigInteger DebitCreditAmount = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, Helper.Concat(_bkeyTag, _bDebitCreditAmount)).AsBigInteger(); //NPCLevel2GGet_cs.txt
                BigInteger Balance = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, Helper.Concat(_bkeyTag, _bBalance)).AsBigInteger(); //NPCLevel2GGet_cs.txt
                e._timestamp = Timestamp; e._decription = Decription; e._debitCreditAmount = DebitCreditAmount; e._balance = Balance;  // Template: NPCLevel2HGet_cs.txt
                e._state = sta;
                e._state = NeoEntityModel.EntityState.GETTED; /* OVERRIDE */
            }
            if (NeoTrace.RUNTIME) LogExt("Get(bkey).NEP5LedgerEntry", e);
            return e;
        }

        public static NEP5LedgerEntry Get(string key)
        {
            if (key.Length == 0) return Null();

            Neo.SmartContract.Framework.Services.Neo.StorageContext ctx = Neo.SmartContract.Framework.Services.Neo.Storage.CurrentContext;
            string _skeyTag = key + _classKeyTag;

            NEP5LedgerEntry e;
            byte[] bsta = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, _skeyTag + _sSTA);
            if (NeoTrace.RUNTIME) TraceRuntime("Get(skey).NEP5LedgerEntry.bsta", bsta.Length, bsta);
            if (bsta.Length == 0)
            {
                e = NEP5LedgerEntry.Missing();
            }
            else // not MISSING
            {
                int ista = (int)bsta.AsBigInteger();
                NeoEntityModel.EntityState sta = (NeoEntityModel.EntityState)ista;
                e = new NEP5LedgerEntry();

                BigInteger Timestamp = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, _skeyTag + _sTimestamp).AsBigInteger(); //NPCLevel2IGet_cs.txt
                string Decription = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, _skeyTag + _sDecription).AsString(); //NPCLevel2IGet_cs.txt
                BigInteger DebitCreditAmount = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, _skeyTag + _sDebitCreditAmount).AsBigInteger(); //NPCLevel2IGet_cs.txt
                BigInteger Balance = Neo.SmartContract.Framework.Services.Neo.Storage.Get(ctx, _skeyTag + _sBalance).AsBigInteger(); //NPCLevel2IGet_cs.txt
                if (NeoTrace.RUNTIME) TraceRuntime("Get(skey).e._timestamp, e._decription, e._debitCreditAmount, e._balance", e._timestamp, e._decription, e._debitCreditAmount, e._balance); // Template: NPCLevel2Part2_cs.txt
                e._timestamp = Timestamp; e._decription = Decription; e._debitCreditAmount = DebitCreditAmount; e._balance = Balance; 
                e._state = sta;
                e._state = NeoEntityModel.EntityState.GETTED; /* OVERRIDE */
            }
            if (NeoTrace.RUNTIME) LogExt("Get(skey).NEP5LedgerEntry", e);
            return e;
        }
    }
}